### ISSUE TITLE
@mysten/sui.js as a peerDependency?

### ISSUE THREAD

**Issue by aslakhellesoy** [shinamicorp/shinami-typescript-sdk#9]
Hello,

Our project uses the latest version of `@mysten/sui.js` (currently `0.41.1`). However, this library uses `^0.39.0`.

Have you considered moving `@mysten/sui.js` from `dependencies` to `peerDependencies` so that the consumers of this library can decide what version to use?

Thanks!
Aslak

**Comment by jasonxh** [comment]
Thanks for the suggestion. My understanding of peer dependency is that we'll have to be very lenient on the version constraint for `@mysten/sui.js`, because it'll prevent consumers from using an incompatible version of `@mysten/sui.js` altogether. With it being a normal dependency, you should still be able to take a direct dependency on another version of `@mysten/sui.js`, and the one depended on by `shinami` will be nested as a result.

The reason I'm reluctant to relax the version constraint on `@mysten/sui.js` is it often contains breaking changes going from `0.X.Y` to `0.X+1.Y` (compliant with semver), and we want to make sure it's tested to work with `shinami`.

Also, we'll be releasing a new `shinami` version soon that's compatible with `@mysten/sui.js 0.41.1`.

**Comment by aslakhellesoy** [comment]
Here is a bit more context:

We've encapsulated our Sui code in a library with functions like this:

```typescript
export async function mintContractStock(
  client: SuiClient,
  params: MintContractStockParams,
): Promise<SuiTransactionBlockResponse> {
  const { contractId, signerPhrase, receiverAddress, packageId, adminCapId, shares } = params;
  const { keypair } = getSigner(signerPhrase);

  const tx = new TransactionBlock();

  tx.moveCall({
    target: `${packageId}::open_art_market::mint_contract_stock`,
    arguments: [
      tx.object(adminCapId),
      tx.object(contractId),
      tx.pure(shares),
      tx.pure(receiverAddress),
    ],
  });

  return client.signAndExecuteTransactionBlock({
    signer: keypair,
    transactionBlock: tx,
    requestType: "WaitForLocalExecution",
    options: {
      showObjectChanges: true,
      showEffects: true,
    },
  });
}
```

We test this library against a local sui node. In those tests we instantiate a `SuiClient` that uses `localnet`.

In production however, we want to invoke the function with a `SuiClient` instantiated with Shinami's `createSuiClient`.
TypeScript isn't happy about this, since our library expects a different version of `SuiClient` than the one provided by Shinami.

If you need to be strict with the version to ensure compatibility, then perhaps you should define it as an exact version number rather than a `^` one?

And to work around our issue, maybe we can use `peerDependency` for the Sui client in our library instead?

**Comment by jasonxh** [comment]
Thanks for the context.

> If you need to be strict with the version to ensure compatibility, then perhaps you should define it as an exact version number rather than a ^ one?

Using `^0.X.Y` would still allow `Y` to increase, which should only contain compatible changes.

> And to work around our issue, maybe we can use peerDependency for the Sui client in our library instead?

I don't think this would help.

 In the example you gave, what you can probably do is to modify your function signature to accept both versions of `SuiClient`, e.g.

```typescript
import { SuiClient, SuiTransactionBlockResponse } from "@mysten/sui.js/client";
import { SuiClient as ShinamiSuiClient } from "shinami/node_modules/@mysten/sui.js/client";

export async function mintContractStock(
  client: SuiClient | ShinamiSuiClient,
  params: MintContractStockParams,
): Promise<SuiTransactionBlockResponse> {
  ...
}
```

This would rely on the compiler to verify that all types involved are compatible across both versions. Also, if you are only using `createSuiClient` from `shinami`, it's probably easier to copy [that logic](https://github.com/shinamicorp/shinami-typescript-sdk/blob/2ddfc979a309e6a3ec030c3a9435c0f6fde3751e/src/sui.ts#L45-L59) from `shinami` sdk and create it yourself. That said, we would also like to keep `shinami` up-to-date on `@mysten/sui.js` versions, so if you don't absolutely require the latest version, you could give us a nudge and we'll try to release a new version as well.

**Comment by jasonxh** [comment]
@aslakhellesoy we just released `shinami@0.2.0` with the updated dependency. Please give it a try and let us know how it goes.

**Comment by jasonxh** [comment]
Closing this issue now. Feel free to reopen it with new information / suggestions.
